students/carlos_novoa/lesson05/mailroom.py

```diff
@@ -0,0 +1,252 @@
+            '        Thank you for your very kind donations totalling {}.\n\n'
+            '        It will be put to very good use.\n\n'
+            '               Sincerely,\n'
+            '                  -The Team')
+    body = str2 if total else str1
+    print(body.format(donor_name, donation))
+
+
+def generate_letters():
+    cwd = os.getcwd()
+    date = now.strftime('%Y-%m-%d')
+    path = cwd + '/letters/'
+    ext = '.txt'
+    for key, val in _donors.items():
+        total_donations = sum(val)
+        name = key.split(' ')
+        file_path = "{}{}_{}_{}{}".format(path, date, name[0], name[1], ext)
+        donations = format_currency_str(total_donations)
+        with open(file_path, 'w') as letter:
+            text = ('Dear {} {},\n\n'
+                    '        Thank you for your very kind donation of {}.\n\n'
+                    '        It will be put to very good use.\n\n'
+                    '               Sincerely,\n'
+                    '                  -The Team')
+            body = text.format(name[0], name[1], donations)
+            letter.write(body)
+    print('Letters created.')
+
+
+def create_report():
+    sorted_donors = sort_donors()
+    rows = get_donor_summary(sorted_donors)
+    print_report(rows)
+
+
+def sort_donors():
+    sorted_donors = sorted(_donors.items(), key=lambda x: x[1], reverse=True)
+    return sorted_donors
+
+
+def get_donor_summary(donors):
+    """ pass summary list of strings that is ready for parsing"""
+    summary = []
+    for name, donations in _donors.items():
+        total = float(sum(donations))
+        number = len(donations)
+        str_total = format_currency_str(total)
+        str_number = str(len(donations))
+        str_average = format_currency_str(total / max(number, 1))
+        summary.append([name, str_total, str_number, str_average])
+    return summary
+
+
+def print_report(rows):
+    # table heading
+    h = ['Donor Name', 'Total Given', 'Num Gifts', 'Average Gift']
+    hs = ' | '
+    hf = '{0:<25}{1}{2}{3}{4}{5}{6}'
+    print(hf.format(h[0], hs, h[1], hs, h[2], hs, h[3]))
+    # table rows
+    for r in rows:
+        name = "{}".format(r[0])
+        f0 = '{0:<' + str(max(len(name), 25)) + '}'
+        f2 = '{2:>' + str(max(len(r[1]), len(h[1]))) + '}'
+        f4 = '{4:>' + str(max(len(r[2]), len(h[2]))) + '}'
+        f6 = '{6:>' + str(max(len(r[3]), len(h[3]))) + '}'
+        rf = f0 + '{1}' + f2 + '{3}' + f4 + '{5}' + f6
+        args = [name, '  $', r[1], ' | ', r[2], '  $', r[3]]
+        print(rf.format(*args))
+
+
+def menu_selection(prompt, dispatch):
+    while True:
+        r = input(prompt)
+        if r not in dispatch:
+            print('Please choose a valid menu option.')
+            continue
+        if dispatch[r]() == "break":
+            break
+
+
+def main_menu():
+    menu_selection(main_prompt, main_dispatch)
+
+
+def donors_sub_menu():
+    menu_selection(donors_prompt, donors_dispatch)
```

Comment:
it would be good to move vars like `donors_prompt` and `donors_dispatch` into this function since they are only used here and more readable.